File: apps/api/app/schemas/us_market.py
"""美股榜单 / 行业·中概板块 契约(0023 阶段③ · 3.3)。

数据源:yfinance 批量拉策展池(决策⑥)· 榜单 = 策展池内排行(非全市场 · 诚实标注)。
成交额为美元估算(close × volume · yfinance 无直接美元成交额)。
frozen + extra=forbid · ts 全 tz-aware UTC。红线:只读行情。
"""

from __future__ import annotations

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field


class UsSpotRow(BaseModel):
    """美股策展池个股快照单行。"""

    model_config = ConfigDict(extra="forbid", frozen=True)

    symbol: str = Field(min_length=1, description="yfinance 代码 · 如 'AAPL'")
    name: str = Field(min_length=1, description="中文名")
    sector: str = Field(min_length=1, description="板块(11 GICS 行业 / 中概股)")
    last_price: float = Field(ge=0)
    change_pct: float = Field(description="涨跌幅 %(可负)")
    amount: float = Field(ge=0, description="成交额(美元估 · close × volume)")
    volume: float = Field(ge=0, description="成交量(股)")


class UsSector(BaseModel):
    """美股板块聚合(策展池内 · 行业 / 中概股)。"""

    model_config = ConfigDict(extra="forbid", frozen=True)

    name: str = Field(min_length=1, description="板块名(科技 / 金融 / 中概股 …)")
    change_pct: float = Field(description="板块涨跌幅 %(成分等权均值 · 可负)")
    stock_count: int = Field(ge=0, description="池内成分股数")
    total_amount: float = Field(ge=0, description="板块成交额(美元估)")


class UsBoardResponse(BaseModel):
    """`GET /api/v1/us/board` 响应 · 重点关注池排行 + 行业/中概板块。

    榜单为同一份策展池快照的 3 种排序(涨幅/跌幅/成交额)· 前端 Tab 内存切换。
    sectors 含「中概股」板块(即中概股板块)。
    """

    model_config = ConfigDict(extra="forbid", frozen=True)

    data_as_of: AwareDatetime | None = Field(default=None, description="快照最新时间 · 截至")
    pool_size: int = Field(ge=0, description="策展池实际有数据的股票数")
    gainers: list[UsSpotRow] = Field(description="涨幅榜(池内 · change_pct DESC)")
    losers: list[UsSpotRow] = Field(description="跌幅榜(池内 · change_pct ASC)")
    top_amount: list[UsSpotRow] = Field(description="成交额榜(池内 · amount DESC)")
    sectors: list[UsSector] = Field(description="行业板块 + 中概股板块(change_pct DESC)")

File: apps/web/lib/api/us-market.ts
/**
 * 美股榜单 / 行业·中概板块 API client(0023 阶段③ · 3.3)。
 *
 * GET /api/v1/us/board · 全市场涨幅/跌幅/成交额 3 榜 + 行业板块。
 */

import { fetchWithRetry } from './fetch-with-retry'

const API_BASE = process.env.NEXT_PUBLIC_API_URL ?? '/api-proxy'

export interface UsSpotRow {
  symbol: string
  name: string
  sector: string
  last_price: number
  change_pct: number
  amount: number // 美元估
  volume: number
}

export interface UsSector {
  name: string
  change_pct: number
  stock_count: number
  total_amount: number
}

export interface UsBoardResponse {
  data_as_of: string | null
  pool_size: number
  gainers: UsSpotRow[]
  losers: UsSpotRow[]
  top_amount: UsSpotRow[]
  sectors: UsSector[]
}

export class UsMarketApiError extends Error {
  constructor(
    public readonly status: number,
    public readonly detail: string,
  ) {
    super(`UsMarketApi ${status}: ${detail}`)
    this.name = 'UsMarketApiError'
  }
}

async function getJson<T>(path: string, signal?: AbortSignal): Promise<T> {
  const r = await fetchWithRetry(`${API_BASE}${path}`, { signal })
  if (!r.ok) {
    let detail = `HTTP ${r.status}`
    try {
      const body = (await r.json()) as { detail?: unknown }
      if (typeof body.detail === 'string') detail = body.detail
    } catch {
      /* ignore */
    }
    throw new UsMarketApiError(r.status, detail)
  }
  return (await r.json()) as T
}

/**
 * 美股全市场榜单 + 行业板块(一次取齐)。
 */
export function fetchUsBoard(limit = 128, signal?: AbortSignal): Promise<UsBoardResponse> {
  return getJson<UsBoardResponse>(`/api/v1/us/board?limit=${limit}`, signal)
}

export function searchUsSpot(
  query: string,
  limit = 50,
  signal?: AbortSignal,
): Promise<UsSpotRow[]> {
  const params = new URLSearchParams({ q: query, limit: String(limit) })
  return getJson<UsSpotRow[]>(`/api/v1/us/search?${params.toString()}`, signal)
}
